contrib/emicklei/go-restful/restful.go

```diff
@@ -26,51 +26,29 @@ func FilterFunc(configOpts ...Option) restful.FilterFunction {
 	}
 	log.Debug("contrib/emicklei/go-restful: Creating tracing filter: %#v", cfg)
 	return func(req *restful.Request, resp *restful.Response, chain *restful.FilterChain) {
-		opts := []ddtrace.StartSpanOption{
-			tracer.ServiceName(cfg.serviceName),
-			tracer.ResourceName(req.SelectedRoutePath()),
-			tracer.SpanType(ext.SpanTypeWeb),
-			tracer.Tag(ext.HTTPMethod, req.Request.Method),
-			tracer.Tag(ext.HTTPURL, req.Request.URL.Path),
-		}
+		var opts []ddtrace.StartSpanOption
 		if !math.IsNaN(cfg.analyticsRate) {
-			opts = append(opts, tracer.Tag(ext.EventSampleRate, cfg.analyticsRate))
-		}
-		if spanctx, err := tracer.Extract(tracer.HTTPHeadersCarrier(req.Request.Header)); err == nil {
-			opts = append(opts, tracer.ChildOf(spanctx))
+			opts = []ddtrace.StartSpanOption{tracer.Tag(ext.EventSampleRate, cfg.analyticsRate)}
 		}
-		span, ctx := tracer.StartSpanFromContext(req.Request.Context(), "http.request", opts...)
-		defer span.Finish()
+		span, ctx := httptrace.StartRequestSpan(req.Request, cfg.serviceName, req.SelectedRoutePath(), false, opts...)
+		defer func() {
+			httptrace.FinishRequestSpan(span, resp.StatusCode(), tracer.WithError(resp.Error()))
+		}()
 
 		// pass the span through the request context
 		req.Request = req.Request.WithContext(ctx)
-
 		chain.ProcessFilter(req, resp)
-
-		span.SetTag(ext.HTTPCode, strconv.Itoa(resp.StatusCode()))
-		span.SetTag(ext.Error, resp.Error())
 	}
 }
 
 // Filter is deprecated. Please use FilterFunc.
 func Filter(req *restful.Request, resp *restful.Response, chain *restful.FilterChain) {
-	opts := []ddtrace.StartSpanOption{
-		tracer.ResourceName(req.SelectedRoutePath()),
-		tracer.SpanType(ext.SpanTypeWeb),
-		tracer.Tag(ext.HTTPMethod, req.Request.Method),
-		tracer.Tag(ext.HTTPURL, req.Request.URL.Path),
-	}
-	if spanctx, err := tracer.Extract(tracer.HTTPHeadersCarrier(req.Request.Header)); err == nil {
-		opts = append(opts, tracer.ChildOf(spanctx))
-	}
-	span, ctx := tracer.StartSpanFromContext(req.Request.Context(), "http.request", opts...)
-	defer span.Finish()
+	span, ctx := httptrace.StartRequestSpan(req.Request, req.SelectedRoutePath(), req.SelectedRoutePath(), false)
+	defer func() {
+		httptrace.FinishRequestSpan(span, resp.StatusCode(), tracer.WithError(resp.Error()))
+	}()
```

Comment:
```suggestion
	defer httptrace.FinishRequestSpan(span, resp.StatusCode(), tracer.WithError(resp.Error()))
```
Or is it possible that `chain.ProcessFilter` changes the return values of `resp.StatusCode()` and `resp.Error()`. I assume that shouldn't be the case because the response is already final isn't it?

Reply:
The arguments would be immediately taken when the deferred call is being stacked and so you wouldn't get the status code nor the error after the call to the handler but rather before (so 0 and nil).

Reply:
Restored in https://github.com/DataDog/dd-trace-go/pull/1286/commits/7dff5a4b6ea21526f1f9cfa535614fa104de3c67

Reply:
Do you mean that `chain.ProcessFilter` might change the response? I just noticed Katie asked the same question...

Reply:
Yes, that's where the next handler is being called AFAIK, resulting in writing the response status code at some point. So the deferred block allows to get the response status code and error after the next filter has been called, when returning from the function.